src/Core/Services/MultipleMutationInputValidator.cs

```diff
@@ -0,0 +1,608 @@
+
+            // As we iterate through columns which can derived from the current entity, we keep removing them from the
+            // columnsToDerive because we know the column value can be derived.
+            // At the end, if there are still columns yet to be derived, i.e.columnsToDerive.count > 0,
+            // we throw an exception.
+            foreach (string columnToBeDerivedFromEntity in columnsToBeSuppliedToSourceEntity)
+            {
+                if (sourceDefinition.Columns[columnToBeDerivedFromEntity].IsAutoGenerated)
+                {
+                    // The value for an autogenerated column is derivable.
+                    columnsToBeSuppliedToSourceEntity.Remove(columnToBeDerivedFromEntity);
+                }
+                else if (backingColumnData.TryGetValue(columnToBeDerivedFromEntity, out IValueNode? value))
+                {
+                    if (value is null)
+                    {
+                        metadataProvider.TryGetExposedColumnName(entityName, columnToBeDerivedFromEntity, out string? exposedColumnName);
+                        throw new DataApiBuilderException(
+                            message: $"Value cannot be null for referenced field: {exposedColumnName} for entity: {entityName} at level: {nestingLevel}.",
+                            statusCode: HttpStatusCode.BadRequest,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.BadRequest);
+                    }
+
+                    columnsToBeSuppliedToSourceEntity.Remove(columnToBeDerivedFromEntity);
+                }
+            }
+
+            // Dictionary storing the mapping from relationship name to the set of referencing fields for all
+            // the relationships for which the current entity is the referenced entity.
+            Dictionary<string, HashSet<string>> fieldsToSupplyToReferencingEntities = new();
+
+            // Dictionary storing the mapping from relationship name to the set of referenced fields for all
+            // the relationships for which the current entity is the referencing entity.
+            Dictionary<string, HashSet<string>> fieldsToDeriveFromReferencedEntities = new();
+
+            // When the source of a referencing field in current entity is a relationship, the relationship's name is added to the value in
+            // the KV pair of (referencing column, source) in derivableColumnsFromRequestBody with a prefix '$' so that if the relationship name
+            // conflicts with the current entity's name or the parent entity's name, we are able to distinguish
+            // with the help of this identifier. It should be noted that the identifier is not allowed in the names
+            // of entities exposed in DAB.
+            string relationshipSourceIdentifier = "$";
+            // Loop over all the relationship fields input provided for the current entity.
+            foreach (ObjectFieldNode fieldNode in objectFieldNodes)
```

Comment:
No, they cannot be renamed to `relationshipFieldNodes` because we are iterating over all the fields but processing only the `relationshipFields` in the loop. I can, at best rename it to `currentEntityInputFields`.